src/Microsoft.Azure.WebJobs.Host/Executors/JobHostContextFactory.cs

```diff
@@ -27,121 +27,162 @@
 
 namespace Microsoft.Azure.WebJobs.Host.Executors
 {
-    internal class JobHostContextFactory : IJobHostContextFactory
+    internal static class JobHostConfigurationExtensions
     {
-        private readonly IStorageAccountProvider _storageAccountProvider;
-        private readonly IConsoleProvider _consoleProvider;
-        private readonly JobHostConfiguration _config;
-
-        public JobHostContextFactory(IStorageAccountProvider storageAccountProvider, IConsoleProvider consoleProvider, JobHostConfiguration config)
+        // Do full initialization (both static and runtime). 
+        // This can be called multiple times on a config. 
+        public static async Task<JobHostContext> CreateAndLogHostStartedAsync(
+            this JobHostConfiguration config, 
+            JobHost host, 
+            CancellationToken shutdownToken, 
+            CancellationToken cancellationToken)
         {
-            _storageAccountProvider = storageAccountProvider;
-            _consoleProvider = consoleProvider;
-            _config = config;
+            var newServices = config.DoStaticInitialization();
+            JobHostContext context = await newServices.DoRuntimeInitialization(config, host, shutdownToken, cancellationToken);
+
+            return context;
         }
 
-        public async Task<JobHostContext> CreateAndLogHostStartedAsync(JobHost host, CancellationToken shutdownToken, CancellationToken cancellationToken)
+        // Static initialization. Returns a service provider with some new services initialized. 
+        // The new services:
+        // - can retrieve static config like binders and converters; but the listeners haven't yet started.
+        // - can be flowed into the runtime initialization to get a JobHost spun up and running.
+        // This shouldn't be async beccause we shouldn't need to do network calls for static init.        
+        // This can be called multiple times on a config, which is why it returns a new ServiceProviderWrapper
+        // instead of modifying the config.
+        public static ServiceProviderWrapper DoStaticInitialization(this JobHostConfiguration config)
+        {
+            var ctx = new ServiceProviderWrapper(config);
+            StaticInitWorker(ctx, config);        
```

Comment:
The current initialization logic is pretty tangled - I couldn't fully untangle it. What's happening here is:
1. initialization creates some new state (tracked via the wrapper). Specifically, static init must carry over some objects to runtime init. 
2. but both inits also depend directly on properties in JobHostConfig (that aren't surfaced as services).